src/LLVM_intrinsics.jl

```diff
@@ -0,0 +1,530 @@
+            )
+        end
+    end
+end
+
+const BINARY_FLOAT_INTRINSICS = [
+    :minnum
+    :maxnum
+    :minimum
+    :maximum
+    :copysign
+    :pow
+    :floor
+    :ceil
+    :trunc
+    :rint
+    :nearbyint
+    :round
+]
+
+for f in BINARY_FLOAT_INTRINSICS
+    @eval @generated function $(f)(x::T, y::T) where T<:LT{<:FloatingTypes}
+        ff = llvm_name($(QuoteNode(f)), T,)
+        return :(
+            $(Expr(:meta, :inline));
+            ccall($ff, llvmcall, T, (T, T), x, y)
+        )
+    end
+end
+
+# pow, powi
+for (f, c) in [(:pow, FloatingTypes), (:powi, IntegerTypes)]
+    @eval @generated function $(f)(x::T, y::T2) where {T <: LT{<:FloatingTypes}, T2 <: $c}
+        ff = llvm_name($(QuoteNode(f)), T)
+        return :(
+            $(Expr(:meta, :inline));
+            ccall($ff, llvmcall, T, (T, T2), x, y)
+        )
+    end
+end
+
+# Comparisons
+const S_CMP_FLAGS = [:eq ,:ne, :sgt ,:sge ,:slt ,:sle]
+const U_CMP_FLAGS = [:eq ,:ne ,:ugt ,:uge ,:ult ,:ule]
+const FCMP_FLAGS = [:false ,:oeq ,:ogt ,:oge ,:olt ,:ole ,:one ,:ord ,:ueq ,:ugt ,:uge ,:ult ,:ule ,:une ,:uno , :true]
+
+for (f, constraint, flags) in zip(("icmp", "icmp", "fcmp"), (IntTypes, UIntTypes, FloatingTypes), (S_CMP_FLAGS, U_CMP_FLAGS, FCMP_FLAGS))
+    for flag in flags
+        ftot = Symbol(string(f, "_", flag))
+        @eval @generated function $ftot(x::LVec{N, T}, y::LVec{N, T}) where {N, T <: $constraint}
+            fflag = $(QuoteNode(flag))
+            ff = $(QuoteNode(f))
+            s = """
+            %res = $ff $(fflag) <$(N) x $(d[T])> %0, %1
+            %resb = zext <$(N) x i1> %res to <$(N) x i8>
+            ret <$(N) x i8> %resb
+            """
+            return :(
+                $(Expr(:meta, :inline));
+                Base.llvmcall($s, LVec{N, Bool}, Tuple{LVec{N, T}, LVec{N, T}}, x, y)
+            )
+        end
+    end
+end
+
+
+#####################
+# Ternary operators #
+#####################
+
+@generated function select(cond::LVec{N, Bool}, x::LVec{N, T}, y::LVec{N, T}) where {N, T}
+    s = """
+    %cond = trunc <$(N) x i8> %0 to <$(N) x i1>
+    %res = select <$N x i1> %cond, <$N x $(d[T])> %1, <$N x $(d[T])> %2
+    ret <$N x $(d[T])> %res
+    """
+    return :(
+        $(Expr(:meta, :inline));
+        Base.llvmcall($s, LVec{N, T}, Tuple{LVec{N, Bool}, LVec{N, T}, LVec{N, T}}, cond, x, y)
+    )
+end
+
+const MULADD_INTRINSICS = [
+    :fmuladd,
+    :fma,
+]
+
+for f in MULADD_INTRINSICS
+    @eval @generated function $(f)(a::LVec{N, T}, b::LVec{N, T}, c::LVec{N, T}) where {N, T<:FloatingTypes}
+        ff = llvm_name($(QuoteNode(f)), N, T)
+        return :(
+            $(Expr(:meta, :inline));
+            ccall($ff, llvmcall, LVec{N, T}, (LVec{N, T}, LVec{N, T}, LVec{N, T}), a, b, c)
+        )
+    end
+end
+
+
+################
+# Load / store #
+################
+
+# These alignment numbers feels a bit dubious
+n_align(align, N, T) = align ? N * sizeof(T) : sizeof(T)
```

Comment:
You probably want `max(align, Base.datatype_alignment(T)))`

Reply:
As is, `vmovntp` won't actually be generating streaming stores if `N` isn't a power of 2 (e.g., 56 byte alignment). That works.

Reply:
Many CPU architectures are able to access memory more efficiently if the alignment is larger than what is required only for `T`. In the end, this is architecture-specific, but a good guess is that the alignment needs to be as large as the underlying vector type (not element type). Thus `N * sizeof(T)` is what I intended here. The unaligned case should probably be `Base.datatype_alignment(T)` instead of `sizeof(T)`:
```Julia
n_align(align, N, T) = align ? N * sizeof(T) : Base.datatype_alignment(T)
```

Reply:
I was confused here. Somehow I was thinking `Base.datatype_alignment(NTuple{7,Core.VecElement{Float64}})`.

Reply:
That would be a good idea, but I don't think it works:
```Julia
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
16
```
My CPU has AVX2 intrinsics, and I would have expected `64` as result. The reported alignment is for SSE2 intrinsics only.

Reply:
What about `sizeof`? Also, what Julia version are you on?

This computer is also AVX2 only:
```julia
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
64

julia> versioninfo()
Julia Version 1.5.0-DEV.255
Commit dd426ba* (2020-02-09 18:28 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz
```

Reply:
> My CPU has AVX2 intrinsics, and I would have expected 64 as result. The reported alignment is for SSE2 intrinsics only.

Did you try on julia 1.4?  Before that 16 is the right answer... (heap alignment), on 1.4 I think we are finally supporting alignments bigger than the heap in more use-cases.

```
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
64
```
this is on 1.5-dev.236

Reply:
Yes, this is Julia 1.4:
```Julia
julia> Base.datatype_alignment(NTuple{8, Core.VecElement{Float64}})
16

julia> versioninfo()
Julia Version 1.4.0-rc1.0
Commit b0c33b0cf5 (2020-01-23 17:23 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.3.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
Environment:
  JULIA_MPI_PATH = /opt/local
  JULIA_MPI_LIBRARY_PATH = /opt/local/lib/openmpi-mp
```

Yes, an alignment of 2048 might cause problems, as it might waste a lot of memory. This should really be an alignment that leads to efficient memory access. Is there a way to query LLVM for the maximum vector size it will actually use for machine instructions? Or should we use the L1 cache line size to limit this value?

Reply:
https://github.com/JuliaLang/julia/pull/34517 is needed for that alignment to show 64.

Reply:
With `Base.datatype_alignment(T)` working like:

```
julia> Base.datatype_alignment(Vec{8, Float64})
64

julia> Base.datatype_alignment(Vec{16, Float64})
128

julia> Base.datatype_alignment(Vec{32, Float64})
256
```

in the upcoming 1.4, aren't we always in the `Aligned`-mode and the default should be changed from what is currently used in SIMD?

Reply:
Vectorized code often has to interface with other, generic code that is not vectorized, by passing arrays forth and back. Arrays might also be multi-dimensional and will usually not be padded for a vectorized alignment. Thus the general case is that arrays are accessed in a non-aligned mode.

That is, these arrays are declared as arrays of scalars (e.g. `Matrix{Float64}`), and we're reading SIMD vectors from it. This is the signature for `vload` and `vstore`. If the array is of type `Matrix{Vec{8, Float64}}`, then it can make sense to assume it is aligned, depending on Julia's array allocator.

Reply:
Yeah, true.